redisstorage.go

```diff
@@ -65,47 +86,78 @@ func (s *Storage) GetCookieJar() http.CookieJar {
 // SetCookies implements http/CookieJar.SetCookies()
 func (s *Storage) SetCookies(u *url.URL, cookies []*http.Cookie) {
 	// TODO RFC 6265
-	cookieStrings := make([]string, 0, len(cookies))
-	for _, c := range cookies {
-		cookieStrings = append(cookieStrings, c.String())
+
+	// TODO(js) Cookie methods currently have no way to return an error.
+
+	// We need to use a write lock to prevent a race in the db:
+	// if two callers set cookies in a very small window of time,
+	// it is possible to drop the new cookies from one caller
+	// ('last update wins' == best avoided).
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	cookieStr, err := s.Client.Get(s.getCookieID(u.Host)).Result()
+	if err == redis.Nil {
+		cookieStr = ""
+	} else if err != nil {
+		// return nil
+		log.Printf("SetCookies() .Get error %s", err)
+		return
 	}
-	for _, c := range s.Cookies(u) {
-		duplication := false
-		for _, c2 := range cookies {
-			if c2.Name == c.Name {
-				duplication = true
-				break
-			}
-		}
-		if !duplication {
-			cookieStrings = append(cookieStrings, c.String())
+
+	// Merge existing cookies, new cookies have precendence.
+	cnew := make([]*http.Cookie, len(cookies))
+	copy(cnew, cookies)
+	existing := unstringify(cookieStr)
+	for _, c := range existing {
+		if !contains(cnew, c.Name) {
+			cnew = append(cnew, c)
 		}
 	}
-	s.Client.Set(s.getCookieID(u.Host), strings.Join(cookieStrings, "\n"), 0).Err()
+	// return s.Client.Set(s.getCookieID(u.Host), stringify(cnew), 0).Err()
+	err = s.Client.Set(s.getCookieID(u.Host), stringify(cnew), 0).Err()
+	if err != nil {
+		// return nil
+		log.Printf("SetCookies() .Set error %s", err)
+		return
+	}
 }
 
 // Cookies implements http/CookieJar.Cookies()
 func (s *Storage) Cookies(u *url.URL) []*http.Cookie {
 	// TODO RFC 6265
-	cookieStr, err := s.Client.Get(s.getCookieID(u.Host)).Result()
-	if err != nil {
+	// TODO(js) Cookie methods currently have no way to return an error.
+
+	s.mu.RLock()
+	cookiesStr, err := s.Client.Get(s.getCookieID(u.Host)).Result()
+	s.mu.RUnlock()
+	if err == redis.Nil {
+		cookiesStr = ""
+	} else if err != nil {
+		// return nil, err
+		log.Printf("Cookies() .Get error %s", err)
 		return nil
 	}
-	header := http.Header{}
-	for _, cs := range strings.Split(cookieStr, "\n") {
-		header.Add("Cookie", cs)
-	}
-	request := http.Request{Header: header}
-	rc := request.Cookies()
-	cookies := make([]*http.Cookie, 0, len(rc))
+
+	// Parse raw cookies string to []*http.Cookie.
+	cookies := unstringify(cookiesStr)
+
+	// Filter.
 	now := time.Now()
-	for _, c := range rc {
-		if c.RawExpires != "" && !c.Expires.After(now) {
+	cnew := make([]*http.Cookie, 0, len(cookies))
+	for _, c := range cookies {
+		// Drop expired cookies.
+		if c.RawExpires != "" && c.Expires.Before(now) {
 			continue
 		}
-		cookies = append(cookies, c)
+		// Drop secure cookies if not over https.
+		if c.Secure && u.Scheme != "https" {
```

Comment:
Arguably, this might not belong in this pull request?

Reply:
The PR fixes multiple issues, so I think it's ok here.